hadoop-hdds/docs/content/design/upgrade-dev-primer.md

```diff
@@ -0,0 +1,96 @@
+
+## Downgrade
+Downgrade to a lower version is allowed from the pre-finalized state. This involves stopping the component, replacing the artifacts to the older version, and then starting it up again.
+
+# Useful framework tools to use
+
+## LayoutFeature
+    org.apache.hadoop.ozone.om.upgrade.OMLayoutFeature
+    org.apache.hadoop.hdds.upgrade.HDDSLayoutFeature
+Class to add  a new layout feature being brought in. Layout version is typically 1 + last layout feature in that catalog. 
+
+## LayoutVersionManager
+    org.apache.hadoop.ozone.om.upgrade.OMLayoutVersionManager
+    org.apache.hadoop.hdds.upgrade.HDDSLayoutVersionManager
+Every component carries an instance of this interface, which provides APIs to get runtime layout version, and if a feature is allowed based on that or not.
+
+## @DisallowedUntilLayoutVersion Annotation
+Method level annotation used to "disallow" an API if current layout version does not include the associated layout feature. Currently it is added only to the OM module, but can easily be moved down to a common module based on need on the HDDS layer.
+
+## @BelongsToLayoutVersion Annotation
+Annotation to mark an OM request class that it belongs to a specific Layout Version. Until that version is available post finalize, this request will not be supported. A newer version of an existing OM request can be created (by inheritance or a fully new class) and marked with a newer layout version. Until finalizing this layout version, the older request class is used. Post finalizing, the newer version of the request class is used.
+
+## LayoutVersionInstanceFactory<T>
+Generic factory which stores different instances of Type 'T' sharded by a key & version. A single key can be associated with different versions of 'T'.
+
+### Why does this class exist?
+A typical use case during upgrade is to have multiple versions of a class / method / object and choose them based  on the current layout version at runtime. Before finalizing, an older version is typically needed, and after finalization, a newer version is needed. This class serves this purpose in a generic way. For example, we can create a Factory to create multiple versions of OMRequests sharded by Request Type & Layout Version Supported.
+
+## Upgrade Action (UpgradeActionOm & UpgradeActionHdds)
+Annotation to specify upgrade action run during specific upgrade phases. Each layout feature can optionally define an upgrade action for every supported phase. These are the supported phases of action callbacks.
+
+#### UNFINALIZED_STATE_VALIDATION
+A ‘validation’ action run every time a component is started up with this layout feature being unfinalized.
+
+#### ON_FIRST_UPGRADE_START
```

Comment:
Are these actions expected to be idempotent, what is the behavior on failure of this action?

Reply:
Failures are not tolerable in any of these actions. In the onFinalize action failure will cause a finalization failure (can either manually finalize or fix the issue in the action). In the other 2 actions, failure will cause the component to have a failed start.

Reply:
Failures are imminent, as long as user can manually proceed and finalize, can you please add that as a note

Reply:
Thanks @swagle, this will be refactored in a follow up patch.